flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamPhysicalMultiJoin.java

```diff
@@ -0,0 +1,221 @@
+import org.apache.calcite.rel.RelWriter;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.hint.RelHint;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.table.planner.utils.ShortcutUtils.unwrapTableConfig;
+
+/**
+ * Stream physical RelNode for a multi-input regular join.
+ *
+ * <p>This node takes multiple inputs and joins them based on a given filter condition. It supports
+ * different join types for each pair of inputs. This node will be translated into a {@link
+ * StreamExecMultiJoin}.
+ */
+public class StreamPhysicalMultiJoin extends AbstractRelNode implements StreamPhysicalRel {
+
+    private List<RelNode> inputs;
+    private final RexNode joinFilter;
+    private final List<JoinRelType> joinTypes;
+    private final List<? extends @Nullable RexNode> joinConditions;
+
+    /**
+     * A map from input index to a list of {@link ConditionAttributeRef}. Each {@link
+     * ConditionAttributeRef} represents a join condition associated with the input.
+     */
+    private final Map<Integer, List<ConditionAttributeRef>> joinAttributeMap;
+
+    private final @Nullable RexNode postJoinFilter;
+    private final List<RelHint> hints;
+
+    public StreamPhysicalMultiJoin(
+            final RelOptCluster cluster,
+            final RelTraitSet traitSet,
+            final List<RelNode> inputs,
+            final RexNode joinFilter,
+            final RelDataType rowType,
+            final List<? extends @Nullable RexNode> joinConditions,
+            final List<JoinRelType> joinTypes,
+            final Map<Integer, List<ConditionAttributeRef>> joinAttributeMap,
+            final @Nullable RexNode postJoinFilter,
+            final List<RelHint> hints) {
+        super(cluster, traitSet);
+        this.inputs = inputs;
+        this.rowType = rowType;
+        this.joinFilter = joinFilter;
+        this.joinTypes = joinTypes;
+        this.joinConditions = joinConditions;
+        this.joinAttributeMap = joinAttributeMap;
+        this.postJoinFilter = postJoinFilter;
+        this.hints = hints;
+    }
+
+    @Override
+    public boolean requireWatermark() {
+        return false;
+    }
+
+    @Override
+    public List<RelNode> getInputs() {
+        return inputs;
+    }
+
+    @Override
+    public void replaceInput(final int ordinalInParent, final RelNode p) {
+        final List<RelNode> newInputs = new ArrayList<>(inputs);
+        newInputs.set(ordinalInParent, p);
+        this.inputs = List.copyOf(newInputs);
+        recomputeDigest();
```

Comment:
2 quetsions
1. are we sure that `ordinalInParent` is always inside boundaries?
2.how likely it could be that value at that index is same as `p`?

Reply:
1. I expect so, I did the same we have for PTFs https://github.com/apache/flink/blob/86ee4b58e3878335d15aeb425025dd8b05973823/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamPhysicalProcessTableFunction.java#L137
2. I don't expect it's likely, since we're replacing it

Reply:
Took a look at it again and 
1. I've added a assertion check since I saw some other places do it
2. As you've mentioned we're not sure what to do if they call with the same `p`, so I think it's safer to let the caller control this